### PR TITLE
ErrorHandling.md: redo Telemetry with events

### PR DESCRIPTION
**Pull request by kainino0x**
Completely separates Telemetry from Fallback/Fatal Errors. Simplifies the log entry API to hopefully include exactly what is necessary (although there is a TODO about whether `object` should be included).

### CODE REVIEW

**Comment by kainino0x** [comment]
Generally, I agree with what you're saying. However, I don't think this is the right mechanism through which to handle it: I'd like to keep this very specifically targeted at telemetry-style error logging.

Aside from maybe the asynchronous nature of #197, I think it more or less provides the primitives an app would need to do this. Here's a sketch of something an app could do:

```ts
class BufferManager {
  device: GPUDevice;
  // hypothetical max heap which is ordered on number
  buffers: MaxHeap<number, GPUBuffer> = new MaxHeap<number, GPUBuffer>();

  async alloc(nice: number, desc: GPUBufferDescriptor): Promise<GPUBuffer> {
    let buffer: GPUBuffer | null = null;
    while (true) {
      let buffer = await this.device.tryCreateBuffer(desc);
      if (buffer) { break; }

      const [nicestNice, nicestBuffer] = this.buffers.pop();
      if (nicestNice <= 0) { throw "OOM"; }
      if (nicestNice <= nice) { return null; }
      nicestBuffer.destroy();
      nicestBuffer.destroyed = true;
    }
    buffer.destroyed = false;
    this.buffers.push(nice, buffer);
    return buffer;
  }
}
```

It's not as ideal as what could be done with a memory heap, but it's something.

**Comment by kenrussell** [comment]
Consider this: OpenGL errors are categorized, and no more than one error of any particular type is reported at a time. Out-of-memory is essentially the only one an app would like to try to recover from. Does it seem a good idea to allow that recovery to be robust - no matter what kind of GPU object was allocated that triggered the OOM?

**Comment by beaufortfrancois** [comment]
Thank you @kainino0x for doing this work! It is much easier to digest with 3 separate PRs. 
Here are my 2 cents below.

TLDR;
- Rename `gpulogentry` to `gpuvalidationerror` for clarity.
- Use an EventHandler for GPUDevice
- Remove `object` and use debug label in newly renamed `message` attribute.

---

```webidl
partial interface GPUDevice : EventTarget {
  attribute EventHandler ongpuvalidationerror;
};
```

`ongpuvalidationerror` is an [EventHandler](https://html.spec.whatwg.org/multipage/webappapis.html#eventhandler) which is called whenever a validation error occurs in the API (including operations on "invalid" WebGPU objects).

```webidl
[
    Constructor(DOMString type, GPUValidationErrorEventInit gpuValidationErrorEventInitDict),
    Exposed=Window
]
interface GPUValidationErrorEvent : Event {
  readonly attribute DOMString message;
};

dictionary GPUValidationErrorEventInit : EventInit {
  required DOMString message;
};
```

Fire an event named "gpuvalidationerror" at GPUDevice using GPUValidationErrorEvent with its message attribute initialized to the validation error message. It should include the debug label of the object if possible.

---

Moreover, in your other PRs, you've include some Javascript examples. And I really liked it. Maybe doing the same here could be nice as well.


```js
const gpuAdapter = await gpu.requestAdapter({ /* options */ });
const gpuDevice = await gpuAdapter.requestDevice({ /* options */ });

gpuDevice.addEventListener('gpuvalidationerror', function(event) {
  console.log(event.message);
  // TODO: Push logs to remote server for telemetry.
});
```

**Comment by kainino0x** [comment]
Applied your changes, thank you for the input! Not 100% certain that all log entries would be validation errors, but did that rename for now.

Can you check the wording for "fires on the main thread event loop"? Not sure how to word it properly.

File: design/ErrorHandling.md [hunk 1]
```diff
@@ -23,59 +23,60 @@ There are several types of WebGPU calls that get their errors handled differentl
...
-These errors are all delivered through one stream - not directly associated with the objects or operations that produced them.
-They should not be used by applications to recover from expected, recoverable errors.
-Instead, the error log may be used for handling unexpected errors in deployment, for bug reporting and telemetry.
-
-```
-enum GPULogEntryType {
-    "device-lost",
-    "validation-error",
-    "recoverable-out-of-memory",
+## *Telemetry*: Error Logging
```
Comment (beaufortfrancois) [inline design/ErrorHandling.md]: Nit: Validation Error Logging

File: design/ErrorHandling.md [hunk 2]
```diff
@@ -23,59 +23,60 @@ There are several types of WebGPU calls that get their errors handled differentl
...
+They are not directly associated with the objects or operations that produced them.
 
- - `"validation-error"`: validation of some WebGPU call failed - including if an argument was an "invalid object".
-   (Either the application did something wrong, or it chose not to recover from a recoverable error.)
- - `"device-lost"`: the `GPUDevice` cannot be used anymore.
-   This may happen if the device is destroyed by the application, reclaimed by the browser, or something goes terribly wrong.
-   (An application may request a new device, or choose to fallback to other content.)
- - `"out-of-memory"`: an allocation failed because too much memory was used by the application (CPU or GPU).
-   This includes recoverable out of memory errors that aren't opt-ed in to be handled by the application when the resource was created.
+The `"validationerror"` event always fires on the main thread (Window) event loop.
```
Comment (beaufortfrancois) [inline design/ErrorHandling.md]: You may want to use the existing WebGL task source as defined in https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.15 and says something like

> The task source for all tasks queued [HTML] in this section is the WebGL task source.

For info, in Picture-in-Picture, I wrote:

> The task source for all the tasks queued in this specification is the media element event task source of the video element in question.

For more, check out https://cs.chromium.org/chromium/src/third_party/blink/public/platform/task_type.h

File: design/sketch.webidl [hunk 3]
```diff
@@ -25,29 +25,12 @@ dictionary GPUExtent3D {
 // ERROR HANDLING
 // ****************************************************************************
 
-enum GPULogEntryType {
-    "device-lost",
-    "validation-error",
-    "recoverable-out-of-memory"
+interface GPULogEntryEvent : Event {
```
Comment (beaufortfrancois) [inline design/sketch.webidl]: It should be replaced with `GPUValidationErrorEvent` right?

File: design/ErrorHandling.md [hunk 4]
```diff
@@ -23,59 +23,60 @@ There are several types of WebGPU calls that get their errors handled differentl
...
 ```
 
-`GPUObjectStatusType` makes a distinction between several error types:
+`WebGPUDevice` `"validationerror"` -> `GPUValidationErrorEvent`:
+Fires when a non-fatal error occurs in the API, i.e. a validation error (including if an "invalid" object is used).
+
+When there is a validation error in the API (including operations on "invalid" WebGPU objects), an error is logged.
+When a validation error is discovered by the WebGPU implementation, it may fire a `"validationerror"` event on the `GPUDevice`.
+These events should not be used by applications to recover from expected, recoverable errors.
+Instead, the error log may be used for handling unexpected errors in deployment, for bug reporting and telemetry.
```
Comment (beaufortfrancois) [inline design/ErrorHandling.md]: Nit: validation errors log

**Comment by kainino0x** [comment]
@beaufortfrancois I have glanced at the Reporting API, but I haven't dug into it enough to understand whether it relates to the kind of reporting we need. Do you know whether we should try to integrate (or merge) this stuff with it?

**Comment by kainino0x** [comment]
Reading a little further it seems like:
- It could be good to surface these errors through it (but it would probably be ideal if we can surface only one error per app error?)
- Not enough adoption to merge, but we can have our error stream feed into it later on

**Comment by grorg** [comment]
Discussed at the [11 Feb 2019 WebGPU Meeting](https://docs.google.com/document/d/1iLZAb6LWiqjoAxsIORXbYdMpspF5a0tIa2uFgUTc_7c/)

**Comment by kainino0x** [comment]
Merged at the meeting, any further comments can be addressed later
